core/trino-main/src/main/java/io/trino/operator/Operator.java

```diff
@@ -92,6 +94,14 @@ default void finishMemoryRevoke()
      */
     boolean isFinished();
 
+    /**
+     * Returns future which will get unblocked when operator becomes finished
+     */
+    default Optional<ListenableFuture<Void>> getFinishedFuture()
```

Comment:
I'm a little worried about expanding the `Operator` interface with a new method. This method doesn't make sense at the `Operator` interface level and will likely create unnecessary confusion, as most of the operators will never have to implement one. 

What do you think about approaching this problem slightly differently. 

Currently a task is transitioned to the `FINISHED` state only when all the drivers are created an complete: https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java#L490

However this requirement is not strongly necessary. In theory a task can be transitioned to the `FINISHED` state as soon as the output drivers are finished. 

The output driver can be easily identified today (see https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/operator/DriverFactory.java#L32, https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/operator/PipelineContext.java#L61)

What do you think about changing the condition and transitioning a task to `FINISHED` state as soon as all the output pipelines / drivers are finished? Once the task is transitioned to `FINISHED` the renaming pipelines / drivers will be unblocked and closed automatically (as it happens today when a task fails).

Reply:
@arhimondr check https://github.com/prestodb/presto/issues/9357.

> SqlTaskExecution.checkTaskCompletion didn't mark the task as finished because the task still has live drivers. (In this particular case, one could argue that the task should be teared down because the only writer to the task output buffer is closed. But in a more general case, this argue may not stand. This is not the root cause, although it could be a valid optimization.)

The fix you suggested wouldn't fix the root cause. Like I could imagine some query plan like:
```
     OutputBuffer
        |
    FinalGlobalAggregation
        |
     Gather
      /  \
  Limit   SubPlanB
     | 
SubPlanA
```
Now could the plan like that be produced in practice

Reply:
> This method doesn't make sense at the Operator interface level and will likely create unnecessary confusion

Why this method doesn't make sense?

Reply:
> Why this method doesn't make sense?

Sorry, i should've phrased it better. It's not like it doesn't make sense, it's just a little confusing. Basically it's an extra method that serves a specific purpose in context of only a specific operator.

Reply:
Discussed offline used `opeator context` instead of new `Operator` interface method